### PR TITLE
Feature/input system

### PR DESCRIPTION
**Pull request by Minoord**
# What changed:
- Created generic input system
- Created swipe input
# How to Test:
- Start up the Test-Input System
- Try swiping on the phone simulator(It can be a bit buggy because it is on pc)

### CODE REVIEW

File: Assets/Scripts/Input/InputSystem.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
+using System.Collections.Generic;
+using System.Diagnostics;
+using UnityEngine.AddressableAssets;
+using UnityEngine.InputSystem;
+
+namespace Input
```
Comment (swzwij) [inline Assets/Scripts/Input/InputSystem.cs]: Namespaces start with the project name and than the folder.

```suggestion
namespace UntitledCube.Input 
```

File: Assets/Scripts/Input/InputSystem.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+using UnityEngine.InputSystem;
+
+namespace Input 
+{
+    public static class InputSystem
+    {
+        private static InputActions _inputAction = new();
+        private static Dictionary<System.Action<InputAction.CallbackContext>, InputAction> _subscribedInputs = new();
+
+        public static InputAction TryGetAction(string actionName)
```
Comment (swzwij) [inline Assets/Scripts/Input/InputSystem.cs]: Missing summary.

File: Assets/Scripts/Input/InputSystem.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            InputAction action = _inputAction.FindAction(actionName);
+            
+            return action;
+        }
+
+        public static void SubscribeToAction(InputAction inputAction, System.Action<InputAction.CallbackContext> function)
```
Comment (swzwij) [inline Assets/Scripts/Input/InputSystem.cs:23]: Missing summary.

File: Assets/Scripts/Input/InputSystem.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        private static InputActions _inputAction = new();
+        private static Dictionary<System.Action<InputAction.CallbackContext>, InputAction> _subscribedInputs = new();
+
+        public static InputAction TryGetAction(string actionName)
+        {
+            InputAction action = _inputAction.FindAction(actionName);
+            
+            return action;
+        }
```
Comment (swzwij) [inline Assets/Scripts/Input/InputSystem.cs]: Can also just be a one-liner

```suggestion
        public static InputAction TryGetAction(string actionName) => _inputAction.FindAction(actionName);
```

File: Assets/Scripts/Input/InputSystem.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+        public static void SubscribeToAction(InputAction inputAction, System.Action<InputAction.CallbackContext> function)
+        {
+            inputAction.Enable();
+            inputAction.performed += function;
+
+            _subscribedInputs.Add(function, inputAction);
+        }
+
+        public static void SubscribeToAction(string inputName, System.Action<InputAction.CallbackContext> function, out InputAction inputAction)
```
Comment (swzwij) [inline Assets/Scripts/Input/InputSystem.cs:37]: Missing summary.

File: Assets/Scripts/Input/InputSystem.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            inputAction = TryGetAction(inputName);
+            inputAction.Enable();
+            inputAction.performed += function;
+
+            _subscribedInputs.Add(function, inputAction);
+        }
+
+        public static void UnsubscribeToAction(InputAction inputAction, System.Action<InputAction.CallbackContext> function)
```
Comment (swzwij) [inline Assets/Scripts/Input/InputSystem.cs:52]: Missing summary.

File: Assets/Scripts/Input/InputSystem.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            inputAction.performed -= function;
+
+            _subscribedInputs.Remove(function, out inputAction);
+        }
+
+        public static void UnsubscribeToAction(string inputName, System.Action<InputAction.CallbackContext> function)
```
Comment (swzwij) [inline Assets/Scripts/Input/InputSystem.cs:64]: Missing summary.

File: Assets/Scripts/Input/InputSystem.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            InputAction inputAction = TryGetAction(inputName);
+            inputAction.performed -= function;
+
+            _subscribedInputs.Remove(function, out inputAction);
+        }
+
+        public static void UnsubscribeToAllActions()
```
Comment (swzwij) [inline Assets/Scripts/Input/InputSystem.cs:75]: Missing summary.

File: Assets/Scripts/Input/InputSystem.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            foreach (KeyValuePair<System.Action<InputAction.CallbackContext>, InputAction> input in _subscribedInputs)
+            {
+                input.Value.performed -= input.Key;
+            }
```
Comment (swzwij) [inline Assets/Scripts/Input/InputSystem.cs]: Can be a one-liner

```suggestion
            foreach (KeyValuePair<System.Action<InputAction.CallbackContext>, InputAction> input in _subscribedInputs)
                input.Value.performed -= input.Key;
```

File: Assets/Scripts/Input/SwipeInteraction.cs [hunk 10]
```diff
@@ -0,0 +1,76 @@
+using System.Collections;
+using System.Collections.Generic;
+using Unity.VisualScripting;
+using UnityEngine;
+using UnityEngine.InputSystem;
+
+namespace Input
```
Comment (swzwij) [inline Assets/Scripts/Input/SwipeInteraction.cs]: Namespaces start with the project name and than the folder.

```suggestion
namespace UntitledCube.Input
```

File: Assets/Scripts/Input/SwipeInteraction.cs [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+{
+    public class SwipeInteraction : MonoBehaviour
+    {
+        [SerializeField] private Vector2 maximumDirection;
+        [SerializeField] private float minimumMagnitude;
```
Comment (swzwij) [inline Assets/Scripts/Input/SwipeInteraction.cs]:
```suggestion
        [SerializeField] private Vector2 _maximumDirection;
        [SerializeField] private float _minimumMagnitude;
```

File: Assets/Scripts/Input/SwipeInteraction.cs [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+            float newX = Mathf.Abs(_newSwipeDirection.x);
+            float newY = Mathf.Abs(_newSwipeDirection.y);
+
+            float directionX = newX - x; 
+            float directionY = newY - y; 
+
+            if(directionX > directionY && _isGoingSideways != 1)
+            {
+                  
+            }*/
```
Comment (swzwij) [inline Assets/Scripts/Input/SwipeInteraction.cs]: Remove commeted code.

File: Assets/Scripts/Input/SwipeInteraction.cs [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+                int direction = _swipeDirection.x > 0 ? 1 : -1;
+                //rotate sideways
+                print("sideways");
+            }
+           else if(x < y)
+            {
+                int direction = _swipeDirection.y > 0 ? 1 : -1;
+                //rotate upwards
+                print("upwards");
+            }
```
Comment (swzwij) [inline Assets/Scripts/Input/SwipeInteraction.cs]:
```suggestion
           if(x > y)
           {
                int direction = _swipeDirection.x > 0 ? 1 : -1;
                //rotate sideways
                print("sideways");
           }
           else if(x < y)
           {
                int direction = _swipeDirection.y > 0 ? 1 : -1;
                //rotate upwards
                print("upwards");
           }
```

**Review by swzwij — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Input/InputSystem.cs (hunk 1 above)
Comment (Minoord) [reply Assets/Scripts/Input/InputSystem.cs]: Changed the namespace

File: Assets/Scripts/Input/InputSystem.cs (hunk 2 above)
Comment (Minoord) [reply Assets/Scripts/Input/InputSystem.cs]: Added it!

File: Assets/Scripts/Input/InputSystem.cs (hunk 3 above)
Comment (Minoord) [reply Assets/Scripts/Input/InputSystem.cs:23]: Added it!

File: Assets/Scripts/Input/InputSystem.cs (hunk 4 above)
Comment (Minoord) [reply Assets/Scripts/Input/InputSystem.cs]: Did that

File: Assets/Scripts/Input/InputSystem.cs (hunk 5 above)
Comment (Minoord) [reply Assets/Scripts/Input/InputSystem.cs:37]: Added it!

File: Assets/Scripts/Input/InputSystem.cs (hunk 9 above)
Comment (Minoord) [reply Assets/Scripts/Input/InputSystem.cs]: Changed it

File: Assets/Scripts/Input/SwipeInteraction.cs (hunk 10 above)
Comment (Minoord) [reply Assets/Scripts/Input/SwipeInteraction.cs]: Changed it!

File: Assets/Scripts/Input/SwipeInteraction.cs (hunk 11 above)
Comment (Minoord) [reply Assets/Scripts/Input/SwipeInteraction.cs]: They are gone

File: Assets/Scripts/Input/SwipeInteraction.cs (hunk 12 above)
Comment (Minoord) [reply Assets/Scripts/Input/SwipeInteraction.cs]: It is gone

File: Assets/Scripts/Input/SwipeInteraction.cs (hunk 13 above)
Comment (Minoord) [reply Assets/Scripts/Input/SwipeInteraction.cs]: Updated it

File: Assets/Scripts/Input/SwipeInteraction.cs [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+        private InputAction _swipeAction;
+
+        private Vector2 _swipePosition;
+        private Vector2 _swipeDirection;
+
+        private void OnEnable()
+        {
+            InputSystem.SubscribeToAction("Swipe", Swipe, out _swipeAction);
+        }
```
Comment (swzwij) [inline Assets/Scripts/Input/SwipeInteraction.cs]:
```suggestion
        private void OnEnable() => InputSystem.SubscribeToAction("Swipe", Swipe, out _swipeAction);
```

File: Assets/Scripts/Input/SwipeInteraction.cs [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+        private void Swipe(InputAction.CallbackContext callbackContext)
+        {
+            _swipePosition = callbackContext.ReadValue<Vector2>();
+
+            Vector2 _newSwipeDirection = CalculateDirection();
+
+            if (_newSwipeDirection == Vector2.zero) return;
```
Comment (swzwij) [inline Assets/Scripts/Input/SwipeInteraction.cs]:
```suggestion
            if (_newSwipeDirection == Vector2.zero) 
                return;
```

File: Assets/Scripts/Input/SwipeInteraction.cs [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+            float y = Mathf.Abs(_swipePosition.y);
+
+            if (x > _moveThreshold && x > y)
+            {
+                return _swipePosition.x > 0 ? Vector2.right : Vector2.left;
+            }
+            else if (y > _moveThreshold && x < y)
+            {
+                return _swipePosition.y > 0 ? Vector2.up : Vector2.down;
+            }
```
Comment (swzwij) [inline Assets/Scripts/Input/SwipeInteraction.cs]:
```suggestion
            if (x > _moveThreshold && x > y)
                return _swipePosition.x > 0 ? Vector2.right : Vector2.left;
            else if (y > _moveThreshold && x < y)
                return _swipePosition.y > 0 ? Vector2.up : Vector2.down;
```

File: Assets/Scripts/Input/SwipeInteraction.cs [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+            return Vector2.zero;
+        }
+
+        private void OnDisable()
+        {
+            InputSystem.UnsubscribeToAction(_swipeAction, Swipe);
+        }
```
Comment (swzwij) [inline Assets/Scripts/Input/SwipeInteraction.cs]:
```suggestion
        private void OnDisable() => InputSystem.UnsubscribeToAction(_swipeAction, Swipe);
```

**Review by swzwij — CHANGES_REQUESTED** [review]

**Review by swzwij — APPROVED** [review]

**Pull request merged**